File: framec_tests/csharp/Basic/BasicTest.cs
using System.IO;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using NUnit.Framework;
#nullable disable
namespace Basic
{
    
class BasicController : Basic
{
    public BasicController() : base() { }

    public new void entered_do(string msg)
    {
        entry_log.Add(msg);
    }

    public new void left_do(string msg)
    {
        exit_log.Add(msg);
    }

}
[TestFixture]
public class BasicTest
{
    [Test]
    public void InitialEnterEvent()
    {
        BasicController sm = new BasicController();
        CollectionAssert.AreEqual(
            sm.entry_log,
            new List<string> { "S0" },
            "Enter event is sent for entering the initial state on startup"
        );
    }

    [Test]
    public void TransitionEnterEvents()
    {
        BasicController sm = new BasicController();
        sm.entry_log = new List<string> {};
        sm.A();
        sm.B();
        CollectionAssert.AreEqual(
            sm.entry_log,
            new List<string> { "S1", "S0" },
            "Enter events are sent to the new state on transition"
        );
    }

    [Test]
    public void TransitionExitEvents()
    {
        var sm = new BasicController();
        sm.A();
        sm.B();
        CollectionAssert.AreEqual(
            sm.exit_log,
            new List<string> { "S0", "S1" },
            "Exit events are sent to the old state on transition"
        );
    }

    [Test]
    public void CurrentState()
    {
        BasicController sm = new BasicController();
        Assert.AreEqual(sm.state_info(), "0");
        sm.A();
        Assert.AreEqual(sm.state_info(), "1");
        sm.B();
        Assert.AreEqual(sm.state_info(), "0");
    }
}

}

File: framec_tests/java/Basic/Basic_test.java
package framec_tests.java.Basic;
import static org.junit.Assert.assertEquals;

import java.util.ArrayList;

import org.junit.Test;

class BasicController extends Basic {
	BasicController() throws Exception {
	  super();
	}
	protected void entered_do(String msg) {
		this.entry_log.add(msg);

	}

	protected void left_do(String msg){
	  this.exit_log.add(msg);
	  
	}
  
} 
  
public class Basic_test{
	
	@Test
	public void TestIntialEnterEvent() throws Exception{
  	BasicController sm = new BasicController();
	ArrayList<String> expected = new ArrayList<String>();
	expected.add("S0");
	assertEquals(expected, sm.entry_log);
	
	}

	@Test
	public void TestTransitionEnterEvents() throws Exception{
		BasicController sm = new BasicController();
		sm.entry_log = new ArrayList<>();
		sm.A();
		sm.B();
		ArrayList<String> expected = new ArrayList<String>();
		expected.add("S1");
        expected.add("S0");
		assertEquals(expected ,sm.entry_log);     	
	
	}
	@Test
	public void TestTransitionExitEvents() throws Exception{
		BasicController sm = new BasicController();
		sm.A();
		sm.B();

		ArrayList<String> expected = new ArrayList<String>();
		expected.add("S0");
		expected.add("S1");
		
    
		assertEquals(expected ,sm.exit_log);  		
	}

	@Test
	public void TestCurrentState() throws Exception{
		BasicController sm = new BasicController();
		assertEquals("0", sm.state_info());
		sm.A();
		assertEquals("1", sm.state_info());
		sm.B();
		assertEquals("0", sm.state_info());
	}

	
}
